src/ATokenVaultFactory.sol

```diff
@@ -0,0 +1,231 @@
+
+    /**
+     * @dev Struct containing constructor parameters for vault deployment
+     */
+    struct VaultParams {
+        address underlying;
+        uint16 referralCode;
+        IPoolAddressesProvider poolAddressesProvider;
+        address owner;
+        uint256 initialFee;
+        string shareName;
+        string shareSymbol;
+        uint256 initialLockDeposit;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Constructor
+     * @param proxyAdmin The address that will be the admin of all deployed proxies
+     */
+    constructor(address proxyAdmin) {
+        require(proxyAdmin != address(0), "ZERO_ADDRESS_NOT_VALID");
+        PROXY_ADMIN = proxyAdmin;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            EXTERNAL FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Deploy a new ATokenVault with the given parameters
+     * @param params All parameters needed for vault deployment and initialization
+     * @return vault The address of the deployed vault proxy
+     */
+    function deployVault(VaultParams memory params) public returns (address vault) {
+        require(params.underlying != address(0), "ZERO_ADDRESS_NOT_VALID");
+        require(address(params.poolAddressesProvider) != address(0), "ZERO_ADDRESS_NOT_VALID");
+        require(params.owner != address(0), "ZERO_ADDRESS_NOT_VALID");
+        require(params.initialLockDeposit > 0, "ZERO_INITIAL_LOCK_DEPOSIT");
+        require(bytes(params.shareName).length > 0, "EMPTY_SHARE_NAME");
+        require(bytes(params.shareSymbol).length > 0, "EMPTY_SHARE_SYMBOL");
+
+        // Transfer the initial lock deposit from caller to this contract
+        IERC20(params.underlying).transferFrom(
+            msg.sender,
+            address(this),
+            params.initialLockDeposit
+        );
+
+        address implementation = ATokenVaultImplDeploymentLib.deployVaultImpl(
+            params.underlying,
+            params.referralCode,
+            params.poolAddressesProvider
+        );
+
+        vault = address(new TransparentUpgradeableProxy(
+            implementation,
+            PROXY_ADMIN,
+            ""
+        ));
+
+        IERC20(params.underlying).approve(vault, params.initialLockDeposit);
+
+        // Initialize the proxy (this will trigger the token transfer)
+        ATokenVault(vault).initialize(
+            params.owner,
+            params.initialFee,
+            params.shareName,
+            params.shareSymbol,
+            params.initialLockDeposit
+        );
+
+        _allVaults.push(vault);
+        _vaultsByUnderlying[params.underlying].push(vault);
+        _vaultsByDeployer[msg.sender].push(vault);
+        _isVaultDeployed[vault] = true;
```

Comment:
do we need this data & reverse mappings onchain? makes the deployment very expensive; if not imo fine to rely on indexer & just maintain _allVaults

Reply:
cc: @joshstevens19

Reply:
Done at 11e3a10be356a8c6883be4f2108539451fa2ee91